apis/defaults.go

```diff
@@ -4,80 +4,78 @@ import (
-	config.MasterConfiguration.APIVersion = "kubeadm.k8s.io/v1alpha1"
-	config.MasterConfiguration.KubernetesVersion = constants.KubernetesVersion
-	config.MasterConfiguration.NoTaintMaster = true
-	addOrAppend(&config.MasterConfiguration.APIServerExtraArgs, "feature-gates", constants.FeatureGates)
-	addOrAppend(&config.MasterConfiguration.ControllerManagerExtraArgs, "feature-gates", constants.FeatureGates)
-	addOrAppend(&config.MasterConfiguration.SchedulerExtraArgs, "feature-gates", constants.FeatureGates)
 }
 
 // SetInitDynamicDefaults sets defaults derived at runtime
 func SetInitDynamicDefaults(config *InitConfiguration) error {
-	nodeName, err := constants.GetHostnameOverride()
-	if err != nil {
-		return fmt.Errorf("unable to dervice hostname override: %v", err)
+	if err := setAPIBindPort(config); err != nil {
+		return fmt.Errorf("unable to set a default value for masterConfiguration.api.bindPort: %s", err)
+	}
+	if err := setAPIControlPlaneEndpoint(config); err != nil {
+		return fmt.Errorf("unable to set a default value for masterConfiguration.api.controlPlaneEndpoint: %s", err)
+	}
+	if err := setKeepalivedInterface(config); err != nil {
+		return fmt.Errorf("unable to set a default value for vipConfiguraton.networkInterface: %s", err)
 	}
-	config.MasterConfiguration.NodeName = nodeName
 	return nil
 }
 
-// SetJoinDefaults sets defaults on the configuration used by join
-func SetJoinDefaults(config *JoinConfiguration) {
-	SetNetworkingDefaults(&config.Networking)
-}
-
-// SetNetworkingDefaults sets defaults for the network configuration
-func SetNetworkingDefaults(netConfig *Networking) {
-	if netConfig.ServiceSubnet == "" {
-		netConfig.ServiceSubnet = constants.DefaultServiceSubnet
+func setAPIBindPort(config *InitConfiguration) error {
+	p, err := gabs.Consume(config.MasterConfiguration)
+	if err != nil {
+		return fmt.Errorf("unable to parse kubeadm MasterConfiguration: %s", err)
 	}
-	if netConfig.DNSDomain == "" {
-		netConfig.DNSDomain = constants.DefaultDNSDomain
+	if p.ExistsP("api.bindPort") {
+		return nil
 	}
+	log.Infof("Setting masterConfiguration.api.bindPort to %d", constants.DefaultAPIBindPort)
+	_, err = p.SetP(constants.DefaultAPIBindPort, "api.bindPort")
+	if err != nil {
+		return fmt.Errorf("unable to set kubeadm MasterConfiguration.api.bindPort: %s", err)
+	}
+	return nil
 }
 
-// SetMasterConfigurationNetworkingDefaultsWithNetworking sets defaults with
-// values from the top-level network configuration
-func SetMasterConfigurationNetworkingDefaultsWithNetworking(config *InitConfiguration) {
-	if config.MasterConfiguration.Networking.ServiceSubnet == "" {
-		config.MasterConfiguration.Networking.ServiceSubnet = config.Networking.ServiceSubnet
+// setAPIControlPlaneEndpoint sets the API ControlPlaneEndpoint to the VIP IP,
+// // if the VIP IP is defined, but the ControlPlaneEndpoint is not.
+func setAPIControlPlaneEndpoint(config *InitConfiguration) error {
+	if config.VIPConfiguration == nil {
```

Comment:
This could be a bug else where, but I want to point on that when no vip is given, the `VIPConfiguration` in the cctl cluster object will give
```
          vipConfiguration:
            routerID: -1
```
So, `VIPConfiguration` would not be nil here: https://play.golang.org/p/nBOqzTrIO4A

Reply:
Thanks for pointing this out. Yes, there's a bug in cctl; I'll file an issue.

Reply:
Filed https://github.com/platform9/cctl/issues/164